File: 프로그래머스/3/150367. 표현 가능한 이진트리/표현 가능한 이진트리.py
def dfs(start, end, tree):
    if start >= end:
        return tree[start]
    
    mid = (start + end) // 2
    
    left = dfs(start, mid - 1, tree)
    right = dfs(mid + 1, end, tree)
    
    if not left or (left == '1' and tree[mid] == '0'):
        return False

    if not right or (right == '1' and tree[mid] == '0'):
        return False
    
    if left == right == tree[mid] == '0':
        return '0'
    return '1'


def convert_to_full_binary_tree(binary_n):
    value = 1
    while 2 ** value - 1 < len(binary_n):
        value += 1
    
    return binary_n.zfill(2 ** value - 1)


def solution(numbers):
    answer = []
    
    for n in numbers:
        binary_n = bin(n)[2:]
        tree = convert_to_full_binary_tree(binary_n)
        
        answer.append(1 if dfs(0, len(tree) - 1, tree) else 0)
        
    return answer

File: 프로그래머스/3/150367. 표현 가능한 이진트리/표현 가능한 이진트리.java
import java.util.Arrays;

public class Solution {

    private boolean isValid(char[] tree, int start, int end, boolean parentIsDummy) {
        if (start > end) {
            return true;
        }

        int mid = (start + end) >>> 1;
        boolean currentIsOne = tree[mid] == '1';
        
        if (parentIsDummy && currentIsOne) {
            return false;
        }

        boolean currentIsDummy = !currentIsOne;

        return isValid(tree, start, mid - 1, currentIsDummy)
                && isValid(tree, mid + 1, end, currentIsDummy);
    }

    private char[] makeFullBinaryTree(long number) {
        String binary = Long.toBinaryString(number);

        int treeLength = 1;
        while (treeLength < binary.length()) {
            treeLength = treeLength * 2 + 1;
        }

        char[] tree = new char[treeLength];

        Arrays.fill(tree, '0');

        int padding = treeLength - binary.length();
        binary.getChars(0, binary.length(), tree, padding);

        return tree;
    }

    public int[] solution(long[] numbers) {
        int[] answer = new int[numbers.length];

        for (int i = 0; i < numbers.length; i++) {
            char[] tree = makeFullBinaryTree(numbers[i]);
            answer[i] = isValid(tree, 0, tree.length - 1, false) ? 1 : 0;
        }

        return answer;
    }
}